fix UCRT migration

danielnachun opened the pull request.

danielnachun committed:
fix UCRT migration

danielnachun committed:
MNT: Re-rendered with conda-build 25.1.1, conda-smithy 3.45.4, and conda-forge-pinning 2025.02.01.03.32.47

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

danielnachun committed:
bump build

danielnachun committed:
fix Windows build

mfansler commented:
FYI, seeing same error here: https://github.com/conda-forge/r-mongolite-feedstock/pull/30

martin-g commented on .ci_support/win_64_.yaml:
```diff
@@ -3,8 +3,14 @@ channel_sources:
 - win-64
 zlib:
-- '1.2'
+- '1'
```
```suggestion
- '1.2'
```

martin-g committed:
Try with zlib 1.2.*

martin-g commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 25.1.1, conda-smithy 3.45.4, and conda-forge-pinning 2025.02.05.19.16.07

martin-g committed:
Debug the existence and location of zlib.h on Windows

martin-g commented:
```
2025-02-06T07:59:19.0241078Z + LDFLAGS=' -L/d/bld/pigz_1738828527320/_h_env/lib -L/d/bld/pigz_1738828527320/_h_env/Library/lib'
2025-02-06T07:59:19.0241723Z + CFLAGS=' -O3 -I/d/bld/pigz_1738828527320/_h_env/include -I/d/bld/pigz_1738828527320/_h_env/Library/include'
2025-02-06T07:59:19.0242323Z + find /d/bld/pigz_1738828527320/_h_env -name '*zlib*'
2025-02-06T07:59:19.0644492Z /d/bld/pigz_1738828527320/_h_env/conda-meta/libzlib-1.3.1-h2466b09_2.json
2025-02-06T07:59:19.0647212Z /d/bld/pigz_1738828527320/_h_env/conda-meta/zlib-1.3.1-h2466b09_2.json
2025-02-06T07:59:19.0649734Z /d/bld/pigz_1738828527320/_h_env/Library/bin/zlib.dll
2025-02-06T07:59:19.0668216Z + make -j2 CC=x86_64-w64-mingw32-gcc.exe 'LDFLAGS= -L/d/bld/pigz_1738828527320/_h_env/lib -L/d/bld/pigz_1738828527320/_h_env/Library/lib' 'CFLAGS= -O3 -I/d/bld/pigz_1738828527320/_h_env/include -I/d/bld/pigz_1738828527320/_h_env/Library/include'
2025-02-06T07:59:19.0669296Z /d/bld/pigz_1738828527320/_h_env/Library/include/zlib.h
2025-02-06T07:59:19.0671177Z /d/bld/pigz_1738828527320/_h_env/Library/lib/pkgconfig/zlib.pc
2025-02-06T07:59:19.0672117Z /d/bld/pigz_1738828527320/_h_env/Library/lib/zlib.lib
2025-02-06T07:59:19.0674510Z /d/bld/pigz_1738828527320/_h_env/Library/lib/zlibstatic.lib
2025-02-06T07:59:19.0676073Z /d/bld/pigz_1738828527320/_h_env/Library/share/man/man3/zlib.3
2025-02-06T07:59:19.0676733Z /d/bld/pigz_1738828527320/_h_env/zlib.dll
2025-02-06T07:59:19.1088610Z x86_64-w64-mingw32-gcc.exe -O3 -I/d/bld/pigz_1738828527320/_h_env/include -I/d/bld/pigz_1738828527320/_h_env/Library/include   -c -o pigz.o pigz.c
2025-02-06T07:59:19.1351587Z x86_64-w64-mingw32-gcc.exe -O3 -I/d/bld/pigz_1738828527320/_h_env/include -I/d/bld/pigz_1738828527320/_h_env/Library/include   -c -o yarn.o yarn.c
2025-02-06T07:59:19.2941576Z pigz.c:418:10: fatal error: zlib.h: No such file or directory
2025-02-06T07:59:19.2950263Z   418 | #include "zlib.h"       // deflateInit2(), deflateReset(), deflate(),
2025-02-06T07:59:19.2972295Z       |          ^~~~~~~~
2025-02-06T07:59:19.2984106Z compilation terminated.
```

`/d/bld/pigz_1738828527320/_h_env/Library/include/zlib.h` is there ! No idea why it is not seen by GCC.

martin-g committed:
Try with CPATH & C_INCLUDE_PATH

martin-g committed:
Add -L${PREFIX}/Library/bin because zlib.dll is there

```
2025-02-06T08:50:46.5692363Z /d/bld/pigz_1738831592997/_h_env/Library/bin/zlib.dll
2025-02-06T08:50:46.5694348Z /d/bld/pigz_1738831592997/_h_env/Library/include/zlib.h
2025-02-06T08:50:46.5696342Z /d/bld/pigz_1738831592997/_h_env/Library/lib/pkgconfig/zlib.pc
2025-02-06T08:50:46.5697502Z /d/bld/pigz_1738831592997/_h_env/Library/lib/zlib.lib
2025-02-06T08:50:46.5698232Z /d/bld/pigz_1738831592997/_h_env/Library/lib/zlibstatic.lib
2025-02-06T08:50:46.5698569Z /d/bld/pigz_1738831592997/_h_env/Library/share/man/man3/zlib.3
2025-02-06T08:50:46.5698870Z /d/bld/pigz_1738831592997/_h_env/zlib.dll
```

martin-g commented:
`export C_INCLUDE_PATH=${PREFIX}/Library/include" helped with the compilation issue!
Now the problem is at linking time:
```
2025-02-06T09:04:54.4913150Z x86_64-w64-mingw32-gcc.exe -L/d/bld/pigz_1738832450432/_h_env/lib -L/d/bld/pigz_1738832450432/_h_env/Library/lib -L/d/bld/pigz_1738832450432/_h_env/Library/bin -o pigz pigz.o yarn.o try.o deflate.o blocksplitter.o tree.o lz77.o cache.o hash.o util.o squeeze.o katajainen.o symbols.o -lm -lpthread -lz
2025-02-06T09:04:54.5564338Z D:/bld/pigz_1738832450432/_build_env/Library/bin/../lib/gcc/x86_64-w64-mingw32/13.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lz: No such file or directory
```

```
2025-02-06T09:04:51.3473455Z /d/bld/pigz_1738832450432/_h_env/Library/bin/zlib.dll
2025-02-06T09:04:51.3488357Z /d/bld/pigz_1738832450432/_h_env/Library/lib/zlib.lib
2025-02-06T09:04:51.3488929Z /d/bld/pigz_1738832450432/_h_env/Library/lib/zlibstatic.lib
2025-02-06T09:04:51.3490212Z /d/bld/pigz_1738832450432/_h_env/zlib.dll
```

How is this supposed to work ?!?
`-lz` means that the linker will try to look for a file name `libz.dll` but here the file is named `zlib.dll`
I don't have experience with Windows.

martin-g committed:
Debug the contents of Library/lib/pkgconfig/zlib.pc

martin-g committed:
Try with pkg-config

martin-g committed:
Add m2-pkg-config for windows

martin-g committed:
export PKG_CONFIG_PATH

martin-g committed:
Update build.sh

martin-g commented:
All is green now!

The linker works when `... -lz -o pigz ...` and fails with `... -o pigz ... -lz`

martin-g committed:
Remove debug. Use pkg-config for Windows

martin-g commented:
@danielnachun Does it look good to you ? I will merge once approved !

danielnachun commented on conda-forge.yml:
```diff
@@ -9,6 +9,6 @@ github:
-test: native_and_emulated
+  linux_aarch64: azure
+  linux_ppc64le: azure
+test: native
```
Let's change this back to the value here:
```suggestion
test: native_and_emulated
```
I misunderstood a problem I was having elsewhere and now I know this is okay.

danielnachun replied:
We'll also need a rerender after that change.

danielnachun commented:
Aside from the small change I suggested above and a rerender, this looks great.  The link ordering issue is expected now that `--as-needed` is used - unfortunately some packages require more work to fix it, but here `pkg-config` did the trick for us!

martin-g committed:
Run tests for both native and emulated

Co-authored-by: Daniel Nachun <daniel.nachun@gmail.com>

martin-g commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 25.1.1, conda-smithy 3.45.4, and conda-forge-pinning 2025.02.06.20.18.51

danielnachun commented:
@martin-g This is ready to merge!
